### PR TITLE
[image_picker_for_web] Support multiple pick. Store name and other local file properties in XFile.

### PR DESCRIPTION
**Pull request by balvinderz**
Add support for picking multiple image , and store name and other properties in XFile

Fixes https://github.com/flutter/flutter/issues/86525
Fixes https://github.com/flutter/flutter/issues/58764

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [X] I read and followed the [relevant style guides] and ran [the auto-formatter]. (Note that unlike the flutter/flutter repo, the flutter/plugins repo does use `dart format`.)
- [X] I signed the [CLA].
- [X] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [X] I listed at least one issue that this PR fixes in the description above.
- [X] I [updated pubspec.yaml](https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates) with an appropriate new version according to the [pub versioning philosophy].
- [X] I updated CHANGELOG.md to add a description of the change.
- [X] I updated/added relevant documentation (doc comments with `///`).
- [X] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test exempt.
- [X] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[the auto-formatter]: https://github.com/flutter/plugins/blob/master/script/tool/README.md#format-code

### CODE REVIEW

**Comment by balvinderz** [comment]
@ditman  can you please review this

**Comment by ditman** [comment]
Ohhh, thanks @balvinderz! Definitely a very requested feature! Thanks for the contribution, I'll review right away!

File: packages/image_picker/image_picker_for_web/lib/image_picker_for_web.dart [hunk 1]
```diff
@@ -96,6 +97,17 @@ class ImagePickerPlugin extends ImagePickerPlatform {
...
+  @override
+  Future<List<PickedFile>?> pickMultiImage(
+      {double? maxWidth, double? maxHeight, int? imageQuality}) {
+    html.FileUploadInputElement input =
+        createInputElement(_kAcceptImageMimeType, null, multiple: true)
+            as html.FileUploadInputElement;
+    _injectAndActivate(input);
+    return _getSelectedFiles(input);
+  }
+
```
Comment (ditman) [inline packages/image_picker/image_picker_for_web/lib/image_picker_for_web.dart:110]: This is only used by a deprecated method of the latest version of the plugin. Let's not implement it for the web. Please remove this implementation (it is also untested :P)

File: packages/image_picker/image_picker_for_web/lib/image_picker_for_web.dart [hunk 2]
```diff
@@ -178,27 +215,61 @@ class ImagePickerPlugin extends ImagePickerPlatform {
...
+        _completer.completeError(event);
+      }
+    });
+    // Note that we don't bother detaching from these streams, since the
+    // "input" gets re-created in the DOM every time the user needs to
+    // pick a file.
+    return _completer.future;
+  }
+
+  Future<List<PickedFile>> _getSelectedFiles(
```
Comment (ditman) [inline packages/image_picker/image_picker_for_web/lib/image_picker_for_web.dart:262]: (Please, do not add any new methods that return PickedFile)

File: packages/image_picker/image_picker_for_web/lib/image_picker_for_web.dart [hunk 3]
```diff
@@ -216,9 +287,40 @@ class ImagePickerPlugin extends ImagePickerPlatform {
...
+        _completer.completeError(event);
+      }
+    });
+    // Note that we don't bother detaching from these streams, since the
+    // "input" gets re-created in the DOM every time the user needs to
+    // pick a file.
+    return _completer.future;
+  }
+
+  Future<List<XFile>> _getSelectedXFiles(html.FileUploadInputElement input) {
```
Comment (ditman) [inline packages/image_picker/image_picker_for_web/lib/image_picker_for_web.dart:310]: There's a lot of logic duplication in the code:

* `_getSelectedFile` and `_getSelectedFiles`, `_handleOnChangeEvent` and `_handleOnChangeEventForMultipleFiles`, etc...

I think the best way to unify logic is to **assume that the user is always selecting multiple files**, and then in the `getImage` just return `.first` of the array of XFiles.

That way we can reduce the size of this change by about half. (Tests would remain the same).

What do you think?

**Review by ditman — CHANGES_REQUESTED** [review]

**Comment by ditman** [comment]
(Check file_selector_web, btw, there's probably a [good example](https://github.com/flutter/plugins/blob/master/packages/file_selector/file_selector_web/lib/file_selector_web.dart#L35-L36) there!)

File: packages/image_picker/image_picker_for_web/lib/image_picker_for_web.dart (hunk 1 above)
Comment (balvinderz) [reply packages/image_picker/image_picker_for_web/lib/image_picker_for_web.dart]: removed it

File: packages/image_picker/image_picker_for_web/lib/image_picker_for_web.dart (hunk 2 above)
Comment (balvinderz) [reply packages/image_picker/image_picker_for_web/lib/image_picker_for_web.dart]: removed this too .

File: packages/image_picker/image_picker_for_web/lib/image_picker_for_web.dart (hunk 3 above)
Comment (balvinderz) [reply packages/image_picker/image_picker_for_web/lib/image_picker_for_web.dart:238]: i have removed the _getSelectedFiles method and i m now returning List<html.File>? from _handleOnChangeEvent.
_getSelectedFile  will now use the first file from _handleOnChangeEvent.
getFile is now returning the first file from _getSelectedXFiles.

File: packages/image_picker/image_picker_for_web/lib/image_picker_for_web.dart [hunk 4]
```diff
@@ -284,6 +317,13 @@ typedef OverrideExtractFilesFromInputFunction = html.File Function(
   html.Element? input,
 );
 
+/// A function that extracts list of files from the file `input` passed in.
+@visibleForTesting
+typedef OverrideExtractMultipleFilesFromInputFunction = List<html.File>
+    Function(
+  html.Element? input,
+);
```
Comment (ditman) [inline packages/image_picker/image_picker_for_web/lib/image_picker_for_web.dart:325]: Did you do `dart format` before submitting the PR? This looks odd :S

**Comment by ditman** [comment]
Tests are failing with:

```
LateInitializationError: Field 'getMultipleFilesFromInput' has not been initialized.
```

Please **run the tests in your machine** ;)

File: packages/image_picker/image_picker_for_web/lib/image_picker_for_web.dart (hunk 4 above)
Comment (balvinderz) [reply packages/image_picker/image_picker_for_web/lib/image_picker_for_web.dart]: i have done it now

**Comment by ditman** [comment]
This is looking great! I'll give this a quick test in my box and have it ready to merge by EOD. Thanks for the contribution @balvinderz! Keep them coming!

**Comment by ditman** [comment]
@googlebot I consent.

**Review by ditman — APPROVED** [review]
LGTM! I've tested this in Chrome Linux and it seems to work as expected. Has tests for the initialization of the new attributes. 🚀

**Comment by ditman** [comment]
Published: https://pub.dev/packages/image_picker_for_web/changelog#211

**Pull request merged**